sys/riotboot/flashwrite.c

```diff
@@ -73,22 +97,51 @@ int riotboot_flashwrite_putbytes(riotboot_flashwrite_t *state,
 
     while (len) {
         size_t flashpage_pos = state->offset % FLASHPAGE_SIZE;
-        size_t flashpage_avail = FLASHPAGE_SIZE - flashpage_pos;
+        size_t flashwrite_buffer_pos = state->offset % RIOTBOOT_FLASHPAGE_BUFFER_SIZE;
+        size_t flashpage_avail = RIOTBOOT_FLASHPAGE_BUFFER_SIZE - flashwrite_buffer_pos;
 
         size_t to_copy = min(flashpage_avail, len);
 
-        memcpy(state->flashpage_buf + flashpage_pos, bytes, to_copy);
+        if (CONFIG_RIOTBOOT_FLASHWRITE_RAW && flashpage_pos == 0) {
+            /* Erase the next page */
+            state->flashpage++;
+            flashpage_write(state->flashpage, NULL);
+        }
+        if (CONFIG_RIOTBOOT_FLASHWRITE_RAW &&
+                flashwrite_buffer_pos == 0) {
+            memset(state->flashpage_buf, 0, RIOTBOOT_FLASHPAGE_BUFFER_SIZE);
+        };
+
+        memcpy(state->flashpage_buf + flashwrite_buffer_pos, bytes, to_copy);
         flashpage_avail -= to_copy;
 
+
         state->offset += to_copy;
         bytes += to_copy;
         len -= to_copy;
         if ((!flashpage_avail) || (!more)) {
-            if (flashpage_write_and_verify(state->flashpage, state->flashpage_buf) != FLASHPAGE_OK) {
+#if CONFIG_RIOTBOOT_FLASHWRITE_RAW  /* Guards access to state::firstblock_buf */
+            void * addr = flashpage_addr(state->flashpage);
+            if (addr == riotboot_slot_get_hdr(state->target_slot) &&
+                    state->offset == RIOTBOOT_FLASHPAGE_BUFFER_SIZE) {
+                /* Skip flashing the first block, copy it over instead */
+                memcpy(state->firstblock_buf,
```

Comment:
Why can we do this?

Reply:
Because it is flashed later during the `riotboot_flashwrite_finish()` function. Main purpose is to make sure that if a node reboots during the flashing, it doesn't detect a half written image as valid.

Reply:
Ah that makes sense.
Mind to add a comment?

Reply:
added a comment